Removed default encoding in refernce to #716

jainamritanshu opened the pull request:
Sir I have removed the hard coded assignments for encoding and added a field of **encoding**. I haven't edited the existing scripts for the existing data. Should I do them manually? Kindly review the code and tell me any changes needed for the code.
@ethanwhite @henrykironde

henrykironde commented:
Am reviewing it?

jainamritanshu commented:
@henrykironde Sir I am extremely sorry. There was a small error from my side. Do you mind if I make a fresh pull request?

henrykironde commented:
In case you see the test failing, ``continuous-integration/appveyor/pr — AppVeyor build failed
`` ,  ``continuous-integration/travis-ci/pr — The Travis CI build failed
``,
it means that there is something that needs attention.

jainamritanshu commented:
@henrykironde I have made the changes to the file `build/compile.py` by handling the KeyError Exception of `json_object['encoding]`. Kindly review and suggest changes whenever time permits.

henrykironde commented:
@jainamritanshu, we should be able to refactor this to one place.

So basically lets try to take care of the the hard-coded encoding initialization encoding='ISO-8859-1' to two places mainly. One in the main __init__ file and an alternative encoding initialization in the scripts.
instead of having this encoding='ISO-8859-1' or encoding='latin-1' everywhere we can initialize it in the main  `__init__.py` file and just import it.

Right now you are working on the encoding in the scripts (if given), 
This is  my idea you could have another way of looking at this. I am open to discus any issues

note:
In` engines\mysql.py:`  we have an edge case `encoding_lookup = {'ISO-8859-1': 'latin1'}`  where `latin-1` is  only understood as `latin1`

jainamritanshu commented:
@henrykironde Should we set a default value to the encoding in `__init__.py` and then the user can change it if he wants while creating the scripts? I think this is what you suggested above and it seems to be a pretty good idea. So the code may go something like - 

We define `ENCODING = <default value of encoding>` in `__init__.py`
We import this value in `lib/compile.py` and `lib/datapackage.py`. And while making the encoding field in datapackage.py - 
```
if is_empty(clean_input("encoding: ", ignore_empty = True)) == "":
    contents['encoding'] = ENCODING
else:
    contents['encoding'] = clean_input("encoding: ", ignore_empty = True) 
```
The edge case in `engines/mysql.py` will also be covered since we will always have a key `'encoding'` in our .py files which will be the default value/value set by the user and hence we can import the value of `encoding` in the file. We can also handle the exception as for the situation if the user enters an unknown encoding value. I am still unsure for this situation and would like to hear your opinions on this case.

This is just a rough idea of what I can picture in my head. Kindly correct/suggest me any other changes you feel I may have skipped.

henrykironde commented:
Looks like that's the way to go. lets make these changes and see what we can add/remove.

jainamritanshu commented:
@henrykironde I have made a couple of changes. Kindly review the code and suggest some changes so that we could proceed further :)

henrykironde commented:
After making the changes, since all the work is very related, we shall squash all the commits into one. Then rebase to master so that we do not have conflicts. If you have finished with the changes and need help with squashing let me know. I will perform a comprehensive review, to have this merged in 👍

jainamritanshu commented:
@henrykironde I have squashed all the commits. Kindly perform a review of the code and suggest any changes you feel might improve the code/working of this feature. :)

henrykironde commented:
@jainamritanshu, I have made some suggestions on what you have to revisit. Let me know once you are done.

jainamritanshu commented:
@henrykironde I could not find your suggestions. I couldn't find it in the issue #716, please help?

henrykironde commented:
@jainamritanshu  no worries, please check the diff. In the `files changed` tab, you do have comments that need to be handled.
`files changed` tab, is right below the heading `Removed default encoding in refernce to #716`

ethanwhite commented:
@henrykironde - I can't see comments on this either. I've forgotten to actually submit the review (I made all of the individual comments but forgot to click the last button so other folks could see them) a couple of times in the past. Is that maybe what's going on here?

henrykironde commented:
Sorry about that, I thought I was actually just commenting yet in reality git was waiting for me to be submit a review.   :) Thanks @ethanwhite

jainamritanshu commented:
@henrykironde I have refactored the code as you suggested. I apologize for so much trouble caused to you while reviewing and I will surely keep these things in my mind for future contributions. Kindly go through the code I have refactored.

henrykironde commented on lib/compile.py:
```diff
@@ -183,19 +183,9 @@ def compile_json(json_file):
+    encoding = 'utf-8'
+    if encoding in json_object : encoding = json_object['encoding']
+    new_script = open(json_file + '.py', 'w', encoding=encoding)
+    new_script.write('# -*- coding: {} -*-\n'.format(encoding)) #Removing the hardcoded encoding
```
@jainamritanshu, this can go away `#Removing the hardcoded encoding`

henrykironde commented on lib/compile.py:
```diff
@@ -179,7 +183,9 @@ def compile_json(json_file):
+    encoding = 'utf-8'
+    if encoding in json_object : encoding = json_object['encoding']
+    new_script = open(json_file + '.py', 'w', encoding=encoding)
+    new_script.write('# -*- coding: {} -*-\n'.format(encoding)) #Removing the hardcoded encoding
```
#Removing the hardcoded encoding

henrykironde commented on lib/datapackage.py:
```diff
@@ -162,6 +162,11 @@ def create_json():
     contents['retriever'] = "True"
     contents['retriever_minimum_version'] = "2.0.dev"
+    # if the user leaves the encoding field empty, set the value of encoding field to default
+    if is_empty(clean_input("encoding: ", ignore_empty = True)) == "":
```
This needs to be revisited

jainamritanshu commented:
@henrykironde I have refactored the code again as suggested by you. Hoping this to be the last review :sweat_smile:

ethanwhite commented on lib/compile.py:
```diff
@@ -179,7 +183,9 @@ def compile_json(json_file):
+    encoding = 'utf-8'
+    if encoding in json_object : encoding = json_object['encoding']
+    new_script = open(json_file + '.py', 'w', encoding=encoding)
+    new_script.write('# -*- coding: {} -*-\n'.format(encoding)) 
```
This code is actually setting the encoding for the `.py` version of the `.json` file, not the encoding of the data, so I don't think we want changes here (unless I'm missing something).

jainamritanshu replied:
So should we keep to latin-1(hardcoded)?

henrykironde replied:
True what Ethan has said, lets not make changes here

ethanwhite commented on lib/engine.py:
```diff
@@ -687,7 +687,7 @@ def to_csv(self):
             csv_writer = open_csvw(csv_file)
             self.get_cursor()
-            self.set_engine_encoding()
+            self.set_engine_encoding(ENCODING)
```
This would pass the default encoding value, which I think would force all engines to set the encoding to the default. I think we want to be passing the actual encoding value instead.

henrykironde commented on lib/compile.py:
```diff
@@ -180,6 +184,6 @@ def compile_json(json_file):
 
     new_script = open(json_file + '.py', 'w', encoding='utf-8')
-    new_script.write('# -*- coding: latin-1 -*-\n')
+    new_script.write('# -*- latin-1 -*-\n') 
```
Remove white space

henrykironde commented on lib/engine.py:
```diff
@@ -655,7 +655,7 @@ def insert_statement(self, values):
         return insert_stmt
 
-    def set_engine_encoding(self):
+    def set_engine_encoding(self): 
```
Remove white space

henrykironde commented on engines/postgres.py:
```diff
@@ -130,5 +131,9 @@ def get_connection(self):
-        conn.set_client_encoding('Latin1')
+
+        encoding = self.script.encoding
+        encoding_lookup = {'ISO-8859-1': 'LATIN1', 'utf-8': 'UTF8'}
```
Include 'latin-1': 'LATIN1' in encoding_lookup.

Also think about a case where we have an encoding in the script like `EUC_KR`
Its not in `encoding_lookup`.

jainamritanshu replied:
So should I take up all possible values of encoding and make a lookup table?

henrykironde replied:
in case of postgres, which may differ from mysql or any other engine, we have https://www.postgresql.org/docs/current/static/multibyte.html.
We do not want to include all, I think Latin and utf are enough. 
For such cases I think we should include a comment saying that incase of any new encoding, update the encoding_lookup with the server's matching value

jainamritanshu replied:
Sure I would make those changes. I was a little confused about the encoding lookup table as to which encoding will map to which one in postgres and mysql database. Did I made them right?

henrykironde replied:
let me know once all the changes are done, and we are doing some great work 👍

jainamritanshu commented:
@henrykironde the tests are facing problems because there isn't a key in 'encoding' in previous scripts. `self.script.encoding` would return `None`. Should I write an exception case here?

henrykironde commented on engines/mysql.py:
```diff
@@ -92,14 +93,18 @@ def table_exists(self, dbname, tablename):
         """
-        encoding_lookup = {'ISO-8859-1': 'latin1'}
-        db_encoding = encoding_lookup.get(encoding, 'latin1')
+
```
Until line 102, I assume the encoding to be ENCODING

```
if self.script.encoding:
    encoding = self.script.encoding.lower()
else:
    encoding = ENCODING

   
encoding_lookup = {'iso-8859-1': 'latin1',    'latin-1': 'latin1',     'utf-8': 'utf8'}
db_encoding = encoding_lookup.get(encoding, ENCODING )
self.execute("SET NAMES '{0}';".format(db_encoding))


```
or the If could be 

```
if self.script.encoding:
     ENCODING = self.script.encoding.lower()
encoding_lookup = {'iso-8859-1': 'latin1',    'latin-1': 'latin1',     'utf-8': 'utf8'}
db_encoding = encoding_lookup.get(ENCODING )
self.execute("SET NAMES '{0}';".format(db_encoding))
```

jainamritanshu replied:
That is an awesome solution, didn't strike my mind. Thanks :)

ethanwhite replied:
Very nice work on this. One very minor here - all caps is used in Python to indicate constants - i.e. variables that never change value. So if we're going to assign a value to `ENCODING` in this way we should use `encoding` or something else lowercase instead.

jainamritanshu replied:
I have assigned a local variable `encoding` in separate engines which will handle this. Also I had made changes in `__init__.py` which are not being reflected. I know that `__init__.py` makes the directory into packages. So am I missing something, please help? I have figured out the error that `__init__.py` is not compiling to make the directory into package. Am I missing something?

henrykironde commented on engines/postgres.py:
```diff
@@ -122,13 +123,21 @@ def format_insert_value(self, value, datatype):
                              password=self.opts["password"],
                              database=self.opts["database"])
-        conn.set_client_encoding('Latin1')
+        encoding = ENCODING
```
I think to lower is missing
`encoding = ENCODING.lower()`

jainamritanshu commented:
@ethanwhite @henrykironde I think I have made all the changes as per the reviews given by you. If there are any other changes I would start working on them right now.

henrykironde commented:
@jainamritanshu, we have a few left and we shall be good to go.
https://github.com/weecology/retriever/pull/810/files#diff-85c3986de1c23543cf73385ebd2eaa9aR187 remove space.

In tests we use `latin-1`, we can go ahead and import ENCODING

henrykironde commented:
@jainamritanshu, go for it. I will review the changes and we make changes accordingly.

jainamritanshu commented:
@henrykironde I have made the respective changes suggested by you. Kindly review the code so I can make some changes if required.

henrykironde commented:
Squash all commits to one and in the test I think we do not need this part.
```
+encoding_lookup = {'iso-8859-1': 'latin1', 'latin-1': 'latin1', 'utf-8': 'utf8'}
 +db_encoding = encoding_lookup.get(encoding)
```
What do you think?

jainamritanshu commented:
@henrykironde I also don't think as a necessary thing, added just for convenience, if in future we felt to change the encoding it could feature as a much easier task to change it if we had the same conventions everywhere in the code base to lookup for the encoding. I will be squashing the commits and if you still suggest to remove I will make the respective changes.

henrykironde commented:
In tests files, I have mine passing with this 

```
from retriever import ENCODING

encoding = ENCODING.lower()

reload(sys)
if hasattr(sys, 'setdefaultencoding'):
    sys.setdefaultencoding(encoding)
```

jainamritanshu commented:
@henrykironde Sure! I will remove the encoding lookup table in tests asap.

henrykironde commented:
@jainamritanshu As i review and test this could you do a `git commit --amend` to edit the commit message examples [ 1](https://github.com/weecology/retriever/commit/a9208eaec63107e5216543523e4bcded547f231d) [2](https://github.com/weecology/retriever/commit/4702ed05c081af938baa12309f2e8bb3f52964ad) [3](https://github.com/weecology/retriever/commit/7cfa9a72d269630cb3952ed2cfb6d475ca7fe373)

henrykironde commented:
In` __main__.py` encoding needs to be used

henrykironde commented:
Thank you for working on this @jainamritanshu.
